quasar-utils/javac-plugin/src/main/java/net/corda/plugins/javac/quasar/SuspendableChecker.java

```diff
@@ -0,0 +1,207 @@
+    }
+
+    private static class Scanner extends TreeScanner<Void, Void> {
+
+        private final Trees trees;
+        private final CompilationUnitTree compilationUnit;
+        private final SuspendableDatabase suspendableDatabase;
+        private ClassTree classTree = null;
+        private MethodTree methodTree = null;
+        private boolean suspendable = false;
+
+        Scanner(Trees trees, CompilationUnitTree compilationUnit, SuspendableDatabase suspendableDatabase) {
+            this.trees = trees;
+            this.compilationUnit = compilationUnit;
+            this.suspendableDatabase = suspendableDatabase;
+        }
+
+        @Override
+        public Void visitClass(ClassTree classTree, Void aVoid) {
+            this.classTree = classTree;
+            return super.visitClass(classTree, aVoid);
+        }
+
+        @Override
+        public Void visitMethod(MethodTree methodTree, Void aVoid) {
+            this.methodTree = methodTree;
+            suspendable = isSuspendable(methodTree);
+            return super.visitMethod(methodTree, aVoid);
+        }
+
+        private boolean isSuspendable(MethodTree tree) {
+            Symbol sym = ((JCTree.JCMethodDecl) tree).sym;
+            return isSuspendable(sym);
+        }
+
+        private static Optional<Symbol> findOverridden(Symbol.TypeSymbol typeSymbol, Symbol methodSymbol) {
+            return StreamSupport.stream(ASTUtils.memberByName(typeSymbol, methodSymbol.name, s ->
+                        Optional.ofNullable(s.type.asMethodType())
+                                .map(mt -> Objects.equals(methodSymbol.asType().getParameterTypes(), mt.getParameterTypes()))
+                                .orElse(false)
+                ).spliterator(), false).findFirst();
+        }
+
+        private boolean superSuspendable(Symbol symbol) {
+            Symbol.ClassSymbol ownerClass = ((Symbol.ClassSymbol) symbol.owner);
+            Stream<Type> supers = Stream.concat(
+                    Optional.ofNullable(ownerClass.getSuperclass())
+                            .filter(t -> t != Type.noType)
+                            .map(Stream::of)
+                            .orElse(Stream.empty()),
+                    ownerClass.getInterfaces().stream());
+            List<Type> supList = supers.collect(Collectors.toList());
+            return supList.stream().map(type -> findOverridden(type.asElement(), symbol))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .anyMatch(s -> inSuspendablesFile(s) || superSuspendable(s));
+        }
+
+        private boolean inSuspendablesFile(Symbol sym) {
+            MethodSignature signature = MethodSignature.from(sym);
+            return suspendableDatabase.isSuspendable(signature.className, signature.methodName, signature.desc);
+        }
+
+        private boolean isSuspendable(Symbol sym) {
+            Annotation annotation = sym.getAnnotation(Suspendable.class);
```

Comment:
This isn't going to work with a configurable `@Suspendable` annotation. (I'm also worried that it's actually referring to your "doppelganger" of Quasar's `@Suspendable` annotation, and not to Quasar's own one at all. What happens when Quasar is actually present on the compile classpath?

Reply:
I actually haven't tried but I expect it to work fine since either of the two will be used by both (if they are in the compiler classpath, both quasar and this plugin will be loaded by the same classloader, so it won't be possible to have duplicate classes). Either way, I agree it is not particularly elegant and I will remove the quasar class duplicates.

Reply:
> I actually haven't tried but ...

I haven't had much luck finding official documentation for Java compiler plugins, and so naturally I haven't found anything to confirm that a plugin's classes will _always_ be hosted in the application classloader either.

Reply:
The plugins need to be specified from the command line simply with `-Xplugin:PluginFullQualifiedClassName` and they are loaded by `javac` assuming them to be in the compilation classpath. 

This is actually quite annoying because it means that a compiler plugin cannot have any third party library dependency (or at least it should shade all of its own dependencies) since otherwise, if your compiler plugin uses Guava 21, that is shared with the code it is supposed to compile, meaning that either your code is tied to the Guava version of your compiler plugin or the other way around.